### PR TITLE
Glyph Atlas improvements

### PR DESCRIPTION
**Pull request by bhousel**
Improves but does not completely eliminate issues with glyph atlas overflow (#141)

* `GlyphSource` now manages multiple `GlyphAtlases`, 1 per fontstack.
* `GlyphAtlas` GL textures start out 128x128 and grow by power of 2 up to 1024x1024
* `BinPack` now uses shelf best height fit algorithm, more efficient packing for things with similar heights
* Doesn't attempt to remove glyphs when removing tiles (performance boost)

### CODE REVIEW

**Comment by bhousel** [comment]
cc @mourner @ansis @jfirebaugh

**Comment by ansis** [comment]
> Doesn't attempt to remove glyphs when removing tiles (performance boost)

Do glyphs ever get removed? How big is the performance benefit?

**Comment by bhousel** [comment]
> Do glyphs ever get removed? How big is the performance benefit?

They do not get removed - for most languages with only a few dozen glyphs this is probably desired anyway as the overhead of constantly removing and then adding them back is not worth it.  I would like to benchmark this rather than just trusting my own perceived performance.

For Asian character sets, I could see a potential rationale for refcounting the glyphs and removing the ones that aren't visible anymore.  

I have been testing around this location in Southern China: #10.01/22.9452/113.1932  - set pan to maximum to load as many tiles as possible, and scroll around.  It does take a LOT of scrolling to get to 2048x2048, and even textures larger than that are considered safe for WebGL.

To exceed the 1024x1024 limit, we would need to change how texture coordinates are packed in `SymbolBucket` `shaderAttributes`.

**Comment by mikemorris** [comment]
> For Asian character sets, I could see a potential rationale for refcounting the glyphs and removing the ones that aren't visible anymore.

Would it be possible to refcount all glyphs added to the atlas and only cull unreferenced glyphs if the atlas would be in danger of overflowing?

**Comment by mourner** [comment]
Yes, I think MRU-cache-like behavior would be optimal — remove the least used glyphs when atlas gets near overflowing. Might need some algorithmic effort though since you would have to use "holes" of different size in the atlas to place new glyphs.

**Comment by mourner** [comment]
As for the PR — the code changes look great. Rebase this into a few clean commits with passing builds and we can merge — even with the overflowing problem, it's better than the current approach.

**Comment by bhousel** [comment]
> As for the PR — the code changes look great. Rebase this into a few clean commits with passing builds and we can merge — even with the overflowing problem, it's better than the current approach.

Great! I squashed away the middle commits and rebased to master.. I'll open a separate issue to track follow on improvements to this.

**Comment by mourner** [comment]
Looks good to me, @ansis want to give a final look?
Also, do we need a similar native PR before we can merge?

**Comment by bhousel** [comment]
I understand there is a need for resizing of sprite atlas also - want me to open a PR for that?  It would be pretty easy to do (at least up to 1024x1024 dimensions).

**Comment by ansis** [comment]
@lucaswoj do you think have a separate glyphAtlas for each fonstack will cause problems with making text-font a data-driven property?

**Comment by lucaswoj** [comment]
> do you think have a separate glyphAtlas for each fonstack will cause problems with making text-font a data-driven property?

Yes, it probably would. But I don't want to delay landing this bug fix for that reason.

**Comment by mikemorris** [comment]
> @lucaswoj do you think have a separate glyphAtlas for each fonstack will cause problems with making text-font a data-driven property?

This will probably need to change to per-font glyph atlases with https://github.com/mapbox/mapbox-gl/issues/4 if that makes a difference re data-driven styling.

**Comment by lucaswoj** [comment]
Upon further thought, I don't think this will affect data-driven styling. We render each symbol in a separate `draw` call. There will be a performance cost to switching glyph atlases between each `draw` call but we can put off digging into that.

**Comment by mourner** [comment]
But this will change after we do #1898, right?

**Comment by lucaswoj** [comment]
If/when we use `ANGLE_instanced_arrays` to render symbols, there'd still be one `draw` call per fontstack

**Comment by ansis** [comment]
@bhousel looks good to me!

The only thing I'm wondering about is whether "not removing glyphs when tiles are removed" is worth it.(cc @kkaefer ) It would be useful to run a quick benchmark on master and see how much tile removeGlyphs takes up.

Other than that, this should be ready to merge!

**Comment by bhousel** [comment]
> The only thing I'm wondering about is whether "not removing glyphs when tiles are removed" is worth it.(cc @kkaefer ) It would be useful to run a quick benchmark on master and see how much tile removeGlyphs takes up.

I did a quick benchmark of this... It's admittedly not great, but I cycled through a 5 glyph-heavy spots and let the map "quiesce" each time.

#### tl;dr:
* `GlyphAtlas.removeGlyphs` takes up 3.5% of the time spent. (I consider this significant).
* On this branch `GlyphAtlas.addGlyphs` jumps up a bit (I think a lot of that is the console warnings, which would go away once we grow the texture as much as I'd like).
* DrawSymbols is slower too, probably from redundant `glyphAtlas.updateTexture(gl);` calls - these can be  minimized.
* _I do agree that a better solution involves refcounting and replacing unused glyphs in the texture, so I want to do that as a next priority._
* merge if you like it - it's at least not worse than what we have now.


#### Master:
screenshot 2016-01-15 16 05 43

#### This Branch:
screenshot 2016-01-15 16 16 16

File: test/js/symbol/binpack.test.js [hunk 1]
```diff
@@ -4,97 +4,51 @@ var test = require('prova');
...
-            y: 0,
-            w: 10,
-            h: 10
-        };
-        t.deepEqual(allocated, expected, 'packs a 10x10 square');
+
+    t.test('not enough room', function(t) {
+        var bp = new BinPack(10, 10);
+        t.deepEqual(bp.allocate(10, 10), {x: 0, y: 0, w: 10, h: 10 }, 'first 10x10 bin');
+        t.deepEqual(bp.allocate(10, 10), {x: -1, y: -1 }, 'not enough room');
```
Comment (jfirebaugh) [inline test/js/symbol/binpack.test.js]: This can be a followup task, but `allocate` should return `null` or `undefined` rather than (-1, -1) to indicate failure.

File: js/style/style_batch.js [hunk 2]
```diff
@@ -151,7 +151,7 @@ styleBatch.prototype = {
         source.id = id;
         source.style = this._style;
         source.dispatcher = this._style.dispatcher;
-        source.glyphAtlas = this._style.glyphAtlas;
+        // source.glyphAtlas = this._style.glyphAtlas;
```
Comment (jfirebaugh) [inline js/style/style_batch.js]: Delete rather than commenting out if not needed.

**Comment by jfirebaugh** [comment]
> merge if you like it

I like it, let's merge and ticket followups.

**Comment by ansis** [comment]
yeah, looks good! thanks for profiling

**Comment by bhousel** [comment]
Thanks! I'll cleanup & merge tonight..

File: js/style/style_batch.js (hunk 2 above)
Comment (bhousel) [inline js/style/style_batch.js]: Sorry, thought I removed them all!

**Comment by bhousel** [comment]
merged as
9d216c1  Limit glyph atlas texture growth to 1024x1024, remove comment
e55c8c5  Use Shelf Best Height Fit algorithm in BinPack, support texture resize
24b5eea  Now GlyphSource will manage multiple GlyphAtlases (1 per fontstack)

**Comment by ansis** [comment]
\o/
